src/transform.ts

```diff
@@ -8,24 +8,81 @@
  */
 
 import * as R from 'ramda';
-import { Left, Right, Either } from 'monads.either';
+
+/**
+ * This class encapsulates either an object of type L (the 'Left'), or
+ * an object of type R (the 'Right'), but not both at the same time,
+ * and never neither of the two.
+ *
+ * Obviously, this class belongs in some type-level util library,
+ * but until that lib exists, this code is the only place it's used.
+ */
+export class Either<L, R> {
```

Comment:
Dig the monad action, one alternative pattern which would allow for the type system to know whether the `Either` is left or right statically (instead of runtime method checks) is to use a discriminated union [(took potential crack at this here)](http://www.typescriptlang.org/play/index.html#src=%0D%0Atype%20Either%3CL%2C%20R%3E%20%3D%20Left%3CL%3E%20%7C%20Right%3CR%3E%3B%0D%0A%0D%0Ainterface%20Left%3CT%3E%20%7B%0D%0A%20%20%20%20kind%3A%20'left'%3B%0D%0A%20%20%20%20value%3A%20T%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Right%3CT%3E%20%7B%0D%0A%20%20%20%20kind%3A%20'right'%2C%0D%0A%20%20%20%20value%3A%20T%0D%0A%7D%0D%0A%0D%0A%0D%0Afunction%20operate%3CL%2C%20R%3E(either%3A%20Either%3CL%2C%20R%3E)%20%7B%0D%0A%20%20%20%20switch%20(either.kind)%20%7B%0D%0A%20%20%20%20%20%20%20%20case%20'right'%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F**%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%20In%20here%2C%20TS%20knows%20that%20the%20value%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%20is%20of%20type%20Right%3CR%3E%20(try%20hovering)%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20*%2F%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20either.value%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20case%20'left'%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20either.value%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20createLeft%3CT%3E(val%3A%20T)%3A%20Left%3CT%3E%20%7B%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20kind%3A%20'left'%2C%0D%0A%20%20%20%20%20%20%20%20value%3A%20val%0D%0A%20%20%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Aconst%20err%20%3D%20createLeft(new%20Error())%3B%0D%0A%0D%0A%2F**%0D%0A%20*%20maintains%20type%20through%20return%20value%0D%0A%20*%2F%0D%0Aconst%20result%20%3D%20operate(err)%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A)